flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultJobGraphStore.java

```diff
@@ -252,9 +255,11 @@ public void putJobGraph(JobGraph jobGraph) throws Exception {
 
                     synchronized (lock) {
                         verifyIsRunning();
-                        if (addedJobGraphs.contains(jobId)) {
+                        if (addedJobGraphs.contains(jobId)
+                                || locallyReleasedJobGraphs.contains(jobId)) {
```

Comment:
why do we need this? I thought we should always call either local or global cleanup, not both. What am I missing?

Reply:
true - I lacked some bigger picture here. The initial wrong-doing was that I removed the `GloballyCleanableResource` interface from the `JobManagerRunnerRegistry`. That required me to call `localCleanup` before calling the `globalCleanup` in the Dispatcher (because we still want to run the cleanup on the `JobManagerRunnerRegistry` even for the `globalCleanup`). 

Essentially, what we want to do is to make `JobManagerRunnerRegistry` implement `GloballyCleanableResource` as well (even though it doesn't have global resources). That enables the `Dispatcher` to call `localCleanup` and `globalCleanup` separately. 🤦